dipy/denoise/fast_lpca.pyx

```diff
@@ -0,0 +1,255 @@
+                    if k>0:
+                        update_matrix(cov, T[cur_i,j,k-1], 1)# q=(0, 0, 1)
+                        update_vector(mu, S[cur_i,j,k-1], 1)# q=(0, 0, 1)
+
+                    # Add displaced signed corners
+                    if i>=m:
+                        x = I[i-m,j,k]# q=(1, 0, 0)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+                        if j>=m:
+                            x = I[i-m,j-m,k]# q=(1, 1, 0)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                            if k>=m:
+                                x = I[i-m,j-m,k-m]# q=(1, 1, 1)
+                                update_outer_prod(cov, x, -1)
+                                update_vector(mu, x, -1)
+                        if k>=m:
+                            x = I[i-m,j,k-m]# q=(1, 0, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if j>=m:
+                        x = I[i,j-m,k]# q=(0, 1, 0)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+                        if k>=m:
+                            x = I[i,j-m,k-m]# q=(0, 1, 1)
+                            update_outer_prod(cov, x, 1)
+                            update_vector(mu, x, 1)
+                    if k>=m:
+                        x = I[i,j,k-m]# q=(0, 0, 1)
+                        update_outer_prod(cov, x, -1)
+                        update_vector(mu, x, -1)
+
+                    # Save current integrals for future reference
+                    update_matrix(T[cur_i, j, k], cov, 0)
+                    update_vector(S[cur_i, j, k], mu, 0)
+
+                    # Use integral of current rectangle
+                    # Last corner is (i, j, k) => center is (i-radius, j-radius, k-radius)
+                    if (i>=m-1) and (j>=m-1) and (k>=m-1):
+                        # Divide by the number of samples
+                        for ii in range(ndiff):
+                            mu[ii]/= nsamples
+                            for jj in range(ndiff):
+                                cov[ii,jj] /= nsamples
+                        # Subtract self-outer product of mu from cov
+                        update_outer_prod(cov, mu, -1)
+
+                        #Local covariance ready
+                        # update_matrix(out[i-radius, j-radius, k-radius], cov, 0)
+                        # update_vector(meanout[i-radius, j-radius, k-radius], mu, 0)
+
+                    with gil:
+                        # eigen value decomposition
+                        [d,W] = np.linalg.eigh(cov)
+                        
+                        if((i >= radius and i < n0-radius) and (j >= radius and j < n1-radius) and (k >= radius and k < n2-radius)):
```

Comment:
Here, `(i, j, k)` is the last corner of the window. The center is `(i-radius, j-radius, k-radius)`. Since all this processing should be part of the above `if`, we know that the center is valid.